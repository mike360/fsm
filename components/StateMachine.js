import React from "react"
import Button from "./Button"
import useFSM, { states } from "../lib/fsm"

export default function StateMachine({ initialState }) {
  const [state, dispatch] = useFSM(initialState)

  return (
    <div>
      <div className="buttons">
        <Button backgroundColor="#4299E1" onClick={() => dispatch("0")}>
          0
        </Button>
        <Button backgroundColor="#03D65C" onClick={() => dispatch("1")}>
          1
        </Button>
      </div>
      <p data-testid="currentState">Current State: {state}</p>
      <p data-testid="outputValue">Output Value: {states[state].output}</p>
      <style jsx>{`
        p {
          font-size: 1.5rem;
          margin: 2rem;
        }
        .states {
          display: flex;
          flex-direction: row;
          justify-content: space-around;
          margin-top: 2rem;
        }
      `}</style>
    </div>
  )
}
