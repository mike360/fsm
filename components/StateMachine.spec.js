import React from "react"
import "@testing-library/jest-dom/extend-expect"
import { render, fireEvent } from "@testing-library/react"
import StateMachine, { reducer } from "./StateMachine"

describe("State Machine", () => {
  describe("Component", () => {
    const testStateTransition = (initialState, input, expectedState) => {
      const { container, getByText, getByTestId, rerender } = render(
        <StateMachine initialState={initialState} />,
      )
      const Button = getByText(input)
      fireEvent.click(Button)
      expect(getByTestId("currentState").textContent).toMatch(
        new RegExp(expectedState),
      )
    }

    describe("When in S0 state", () => {
      it("first renders with initial state", () => {
        const { container, getByText } = render(<StateMachine />)
        expect(container).toMatchSnapshot()
      })
      it("updates the state to S1 when the 1 button is clicked", () => {
        testStateTransition("S0", "1", /S1/)
      })

      it("stays in S0 state when 0 button is clicked", () => {
        testStateTransition("S0", "0", /S0/)
      })
    })

    describe("When in S1 state", () => {
      it("updates the state to S2 when 0 is clicked", () => {
        testStateTransition("S1", "0", /S2/)
      })

      it("updates the state to S0 when 1 is clicked", () => {
        testStateTransition("S1", "1", /S0/)
      })
    })

    describe("When in S2 state", () => {
      it("updates the state to S1 when 0 is clicked", () => {
        testStateTransition("S2", "0", /S1/)
      })

      it("stays in S2 when 1 is clicked", () => {
        testStateTransition("S2", "1", /S2/)
      })
    })
  })
})
