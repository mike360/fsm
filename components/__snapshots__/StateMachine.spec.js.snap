// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`State Machine Component When in S0 state first renders with initial state 1`] = `
<div>
  <div
    class="jsx-1300285096"
  >
    <div
      class="jsx-1300285096 buttons"
    >
      <button
        class="jsx-4025961682 "
      >
        0
      </button>
      <button
        class="jsx-720380633 "
      >
        1
      </button>
    </div>
    <p
      class="jsx-1300285096"
      data-testid="currentState"
    >
      Current State: 
      S0
    </p>
    <p
      class="jsx-1300285096"
      data-testid="outputValue"
    >
      Output Value: 
      0
    </p>
  </div>
</div>
`;
